README.md
# Buddha's Tradovate Trading Scripts

---

## Upper Indicators

-   Buddha SMA Bands
-   Buddha V-wap

## Lower Indicators

-   Buddha Squeeze
-   Buddha STOCH RSI
-   Buddha Volume

---

Darth Buddha


CHANGELOG.md
# Buddha's Tradovate Trading Scripts

---

## Buddha's SMA Bands

buddhaSmaBands.js

## Buddha's TTM Squeeze

buddhaSqueeze.js

## Buddha's Stoch Rsi

buddhaStochRsi.js

## Buddha's Bulls vs Bears Volume

buddhaVolume.js

## Buddha's Vwap

buddhaVwap.js

### Changelog

#### 0.0.0

Start of something good!

New:

-   buddhaSmaBands.js
-   buddhaSqueeze.js
-   buddhaStochRsi.js
-   buddhaVolume.js
-   buddhaVwap.js

##### Darth Buddha's Tradovate Trading Scripts


scripts/buddhaVwap.js
/*********************************
 * Darth Buddha Tradovate Script *
 * *******************************
 * Buddha V Wap
 * Build : 25
 *********************************/
/*  Constant  Variables  */
const predef = require("../tools/predef");
const meta = require("../tools/meta");
const typicalPrice = require("../tools/typicalPrice");
/*  Initialize  Class  */
class buddhaVwap {
    init() {
        this.cumulativeValue = 0;
        this.cumulativeVolume = 0;
        this.tradeDate = 0;
    }

    map(d) {
        const tradeDate = d.tradeDate();
        if (tradeDate !== this.tradeDate) {
            this.cumulativeVolume = 0;
            this.cumulativeValue = 0;
            this.tradeDate = tradeDate;
        }
        const volumeProfile = d.profile();
        if (volumeProfile && volumeProfile.length) {
            for (let i = 0; i < volumeProfile.length; ++i) {
                const level = volumeProfile[i];
                this.cumulativeVolume += level.vol;
                this.cumulativeValue += level.vol * level.price;
            }
        } else {
            const vol = d.volume();
            this.cumulativeVolume += vol;
            this.cumulativeValue += vol * typicalPrice(d);
        }
        return this.cumulativeValue / this.cumulativeVolume;
    }
}
/*  Export  Module  */
module.exports = {
    name: "buddhaVwap",
    description: "Buddha V Wap",
    calculator: buddhaVwap,
    tags: ["Buddha Indicators"],
    inputType: meta.InputType.BARS,
    schemeStyles: predef.styles.solidLine({
        color: "#FFFFFF",
        lineWidth: 1,
        lineStyle: 4,
    }),
};
/*  END  CODE  */


scripts/buddhaVolume.js
/*********************************
 * Darth Buddha Tradovate Script *
 * *******************************
 * Buddha Volume
 * Build : 25
*********************************/
/*  Constant  Variables  */
const predef = require("../tools/predef");
const meta = require("../tools/meta");
const p = require("../tools/plotting");
const SMA = require("../tools/SMA");

const defaultBearColor      = "#D0021B";
const defaultBullColor      = "#417505";
const defaultVolAvgColor    = "#00FFFF";
const defaultVolAvgPeriod   = 21;
const defaultLineWidthPx    = 3;
/*  Initialize  Class  */
class buddhaVolume {
    init() {
        this.smaAlgo = SMA(this.props.volAvgPeriod);
    }

    map(d) {
        const volume = d.volume();
        const close = d.close();
        const high = d.high();
        const low = d.low();

        const range = high - low;
        const bulls = Math.round(volume * (close - low) / range);
        const bears = Math.round(volume * (high - close) / range);

        // use bid/offer volumes
        //const bulls = d.offerVolume();
        //const bears = d.bidVolume();

        return {
            volume: volume,
            bears: bears,
            bulls: bulls,
            volAvg: this.smaAlgo(volume)
        };
    }
}
/*  Export  Module  */
module.exports = {
    name: "buddhaVolume",
    description: "Buddha Volume",
    calculator: buddhaVolume,
    tags: ["Buddha Indicators"],
    inputType: meta.InputType.BARS,
    areaChoice: meta.AreaChoice.NEW,
    params: {
        volAvgPeriod: predef.paramSpecs.period(defaultVolAvgPeriod),
    },
    plots : {
        volume: {displayOnly:true},
        bears: 'bears',
        bulls: 'bulls',
        volAvg: {title:'VolAvg'},

    },
    plotter: [
        {
            type:'histogram',
            fields: ['volume', 'bears','bulls']
        },
        predef.plotters.singleline("volAvg")
    ],
    schemeStyles: {
        dark: {
            volume: predef.styles.plot({
                color: "#4A4A4A"
            }),
            bears: {color: defaultBearColor},
            bulls: {color: defaultBullColor},
            volAvg: predef.styles.plot({
                color: defaultVolAvgColor,
                lineWidth: 3
            })
        }
    },
};
/*  END  CODE  */


scripts/buddhaSqueeze.js
/*********************************
 * Darth Buddha Tradovate Script *
 * *******************************
 * Buddha Squeeze
 * Build : 25
 *********************************/
/*  Constant  Variables  */
const predef = require("../tools/predef");
const meta = require("../tools/meta");
const EMA = require("../tools/EMA");
const SMA = require("../tools/SMA");
const trueRange = require("../tools/trueRange");
const StdDev = require("../tools/StdDev");
/*  Initialize  Class  */
class buddhaSqueeze {
    init() {
        this.prices = [];
        this.closeEMA = EMA(this.props.period);
        this.closeSMA = SMA(this.props.period);
        this.atrSMA = SMA(this.props.period);
        this.stdDev = StdDev(this.props.period);
    }
    map(d, i, history) {
        const period = this.props.period;
        if (this.prices.length === period) {
            this.prices.shift();
        }

        const close = d.close();
        const high = d.high();
        const low = d.low();

        // Keltner Channel
        const closeSMA = this.closeSMA(close);
        const atr =
            this.props.kcMultiplier *
            this.atrSMA(trueRange(d, history.prior()));
        const kcUpper = closeSMA + atr;
        const kcLower = closeSMA - atr;

        // Bollinger Band
        const stdev = this.stdDev(close);
        const avg = this.stdDev.avg();
        const halfWidth = stdev * this.props.bbMultiplier;
        const bbUpper = avg + halfWidth;
        const bbLower = avg - halfWidth;

        const isSqueeze = bbUpper < kcUpper && bbLower > kcLower;

        const prices = {
            close,
            high,
            low,
        };
        this.prices.push(prices);

        const { highest, lowest } = this.prices.reduce(
            (result, price) => {
                if (price.high > result.highest) {
                    result.highest = price.high;
                }
                if (price.low < result.lowest) {
                    result.lowest = price.low;
                }
                return result;
            },
            { highest: 0, lowest: Infinity }
        );

        const closeEMA = this.closeEMA(close);
        const value = close - ((highest + lowest) / 2 + closeEMA) / 2;
        prices.value = value;

        let sumX = 0;
        let sumY = 0;
        let sumXY = 0;
        let sumXX = 0;
        const n = this.prices.length;
        this.prices.forEach((price, i) => {
            const x = i + 1;
            sumY += price.value;
            sumX += x;
            sumXY += price.value * x;
            sumXX += x * x;
        });
        const slope = (n * sumXY - sumX * sumY) / (n * sumXX - sumX * sumX);
        const intercept = (sumY - slope * sumX) / n;
        const regression = intercept + slope * n;
        const Squeeze = close;
        const last = this.last || regression;
        this.last = regression;

        const {
            squeezeOnColor,
            squeezeOffColor,
            upAdvanceColor,
            upDeclineColor,
            downAdvanceColor,
            downDeclineColor,
            isSqueezeColor,
            colorBars,
        } = this.props;

        const squeezeColor = isSqueeze ? squeezeOnColor : squeezeOffColor;
        const histogramColor =
            regression > 0
                ? regression > last
                    ? upAdvanceColor
                    : upDeclineColor
                : regression < last
                ? downDeclineColor
                : downAdvanceColor;
        const candlestickColor = isSqueeze ? isSqueezeColor : histogramColor;
        const candlestick = colorBars
            ? {
                  color: candlestickColor,
              }
            : undefined;

        return {
            squeeze: 0,
            value: regression,
            style: {
                value: {
                    color: histogramColor,
                },
                squeeze: {
                    color: squeezeColor,
                },
            },
            candlestick,
        };
    }
}
/*  Export  Module  */
module.exports = {
    name: "buddhaSqueeze",
    description: "Buddha Squeeze",
    calculator: buddhaSqueeze,
    tags: ["Buddha Indicators"],
    params: {
        period: predef.paramSpecs.period(21),
        kcMultiplier: predef.paramSpecs.number(1.5, 0.1),
        bbMultiplier: predef.paramSpecs.number(2.0, 0.1),
        colorBars: predef.paramSpecs.bool(false),
        squeezeOffColor: predef.paramSpecs.color("#00000000"),
        squeezeOnColor: predef.paramSpecs.color("#F8E71C"),
        upAdvanceColor: predef.paramSpecs.color("#417505"),
        upDeclineColor: predef.paramSpecs.color("#D0021B"),
        downAdvanceColor: predef.paramSpecs.color("#7ED321"),
        downDeclineColor: predef.paramSpecs.color("#780000"),
        isSqueezeColor: predef.paramSpecs.color("#F8E71C"),
    },
    plots: {},
    plotter: [{ type: "histogram" }, { type: "dots", field: "squeeze" }],
    inputType: meta.InputType.BARS,
    areaChoice: meta.AreaChoice.NEW,
    schemeStyles: {
        dark: {
            value: predef.styles.plot({ color: "#00000000", lineWidth: 0 }),
            squeeze: predef.styles.plot({ color: "#F8E71C", lineWidth: 1 }),
        },
    },
};
/*  END  CODE  */


scripts/buddhaStochRsi.js
/*********************************
 * Darth Buddha Tradovate Script *
 * *******************************
 * Buddha STOCH RSI
 * Build : 25
*********************************/
/*  Constant  Variables  */
const predef = require("./tools/predef");
const meta = require("./tools/meta");
const MovingHigh = require("./tools/MovingHigh");
const MovingLow = require("./tools/MovingLow");
const SMA = require("./tools/SMA");
const EMA = require('./tools/EMA');
const WMA = require('./tools/WMA');
const MMA = require('./tools/MMA');

const ma = {
    sma: SMA,
    ema: EMA,
    wma: WMA,
    mma: MMA
};
/*  Initialize  Class  */
class buddhaStochRsi {
    init() {
        this.highest = MovingHigh(this.props.kPeriod);
        this.lowest = MovingLow(this.props.kPeriod);
        this.kMA = ma[this.props.type](this.props.smoothPeriod);
        this.dMA = ma[this.props.type](this.props.dPeriod);
    }

    map(d) {
        const high = d.high();
        const low = d.low();
        const close = d.close();
        const hh = this.highest(high);
        const ll = this.lowest(low);
        const c1 = close - ll;
        const c2 = hh - ll;
        const fastK = c2 !== 0 ? c1 / c2 * 100 : 0;
        const K = this.kMA(fastK);
        const D = this.dMA(K);
        return { 
            K, 
            D, 
            overBought: this.props.overBought,
            overSold: this.props.overSold,
            lineTop: this.props.lineTop,
            lineBottom: this.props.lineBottom,
        };
    }

    filter(d) {
        return predef.filters.isNumber(d.D);
    }
}
/*  Export  Module  */
module.exports = {
    name: "buddhaStochRsi",
    title: "Buddha Stoch RSI",
    description: "Buddha Stoch RSI",
    calculator: buddhaStochRsi,
    tags: ['Buddha Indicators'],
    params: {
        kPeriod: predef.paramSpecs.period(14),
        dPeriod: predef.paramSpecs.period(3),
        smoothPeriod: predef.paramSpecs.period(1),
        overBought: predef.paramSpecs.period(80),
        overSold: predef.paramSpecs.period(20),
        lineTop: predef.paramSpecs.period(110),
        lineBottom: predef.paramSpecs.period(-10),
        type: predef.paramSpecs.enum({
            sma: 'Simple',
            ema: 'Exponential',
            wma: 'Weighted',
            mma: 'Modified'
        }, 'sma')
    },
    validate(obj) {
        if (obj.period < 1) {
            return meta.error("kPeriod", "K Period should be a positive number");
        }
        if (obj.period < 1) {
            return meta.error("dPeriod", "D Period should be a positive number");
        }
        if (obj.smoothPeriod < 2) {
            return meta.error("smoothPeriod", "Smooth period should be greater than 1");
        }
        return undefined;
    },
    inputType: meta.InputType.BARS,
    areaChoice: meta.AreaChoice.NEW,
    plots: {
        K: { title: "%K" },
        D: { title: "%D" },
        overBought: { title: "Over Bought" },
        overSold: { title: "Over Sold" },
        lineTop: { title: "Top Line" },
        lineBottom: { title: "Bottom Line"},
    },
    schemeStyles: {
        dark: {
            K: predef.styles.plot({
                color: "green",
                lineWidth: 2
            }),
            D: predef.styles.plot({
                color: "red",
                lineWidth: 1
            }),
            overBought: predef.styles.plot({
                color: "white",
                lineStyle: 3
            }),
            overSold: predef.styles.plot({
                color: "white",
                lineStyle: 3
            }),
            lineTop: predef.styles.plot({
                color: "#000000",
                lineStyle: 1
            }),
            lineBottom: predef.styles.plot({
                color: "#000000",
                lineStyle: 1
            })
        }
    }
};
/*  END  CODE  */

scripts/buddhaSmaBands.js
/*********************************
 * Darth Buddha Tradovate Script *
 * *******************************
 * Buddha SMA Bands
 * Build : 25
 *********************************/
/*  Constant  Variables  */
const predef = require("../tools/predef");
const SMA = require("../tools/SMA");
const { du, op, px, min, max } = require("../tools/graphics");
/*  Initialize  Class  */
class buddhaSmaBands {
    init() {
        this.newSma01 = SMA(this.props.period20);
        this.oldSma01 = SMA(this.props.period20);

        this.newSma02 = SMA(this.props.period21);
        this.oldSma02 = SMA(this.props.period21);

        this.newSma03 = SMA(this.props.period50);
        this.oldSma03 = SMA(this.props.period50);

        this.newSma04 = SMA(this.props.period55);
        this.oldSma04 = SMA(this.props.period55);

        this.newSma05 = SMA(this.props.period200);
        this.oldSma05 = SMA(this.props.period200);

        this.newSma06 = SMA(this.props.period233);
        this.oldSma06 = SMA(this.props.period233);

        this.newSma07 = SMA(this.props.period365);
        this.oldSma07 = SMA(this.props.period365);

        this.newSma08 = SMA(this.props.period377);
        this.oldSma08 = SMA(this.props.period377);
    }

    map(d, i, history) {
        const value = d.value();

        if (i < 1) return {};
        const last = history.back(1);

        const oldSma20 = this.oldSma01(last.value());
        const oldSma21 = this.oldSma02(last.value());
        const oldSma50 = this.oldSma03(last.value());
        const oldSma55 = this.oldSma04(last.value());
        const oldSma200 = this.oldSma05(last.value());
        const oldSma233 = this.oldSma06(last.value());
        const oldSma365 = this.oldSma07(last.value());
        const oldSma377 = this.oldSma08(last.value());

        const newSma20 = this.newSma01(d.value());
        const newSma21 = this.newSma02(d.value());
        const newSma50 = this.newSma03(d.value());
        const newSma55 = this.newSma04(d.value());
        const newSma200 = this.newSma05(d.value());
        const newSma233 = this.newSma06(d.value());
        const newSma365 = this.newSma07(d.value());
        const newSma377 = this.newSma08(d.value());

        return {
            sma20: newSma20,
            sma21: newSma21,
            sma50: newSma50,
            sma55: newSma55,
            sma200: newSma200,
            sma233: newSma233,
            sma365: newSma365,
            sma377: newSma377,
            graphics: {
                items: [
                    {
                        tag: "Shapes",
                        key: "sma20x21Shading",
                        primitives: [
                            {
                                tag: "Polygon",
                                points: [
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma21),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma21),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma20),
                                    },
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma20),
                                    },
                                ],
                            },
                        ],
                        fillStyle: {
                            opacity: this.props.bandOpacity,
                            color:
                                newSma20 > newSma21
                                    ? this.props.sma20x21BandUp
                                    : this.props.sma20x21BandDown,
                        },
                    },
                    {
                        tag: "Shapes",
                        key: "sma50x55Shading",
                        primitives: [
                            {
                                tag: "Polygon",
                                points: [
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma55),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma55),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma50),
                                    },
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma50),
                                    },
                                ],
                            },
                        ],
                        fillStyle: {
                            opacity: this.props.bandOpacity,
                            color:
                                newSma50 > newSma55
                                    ? this.props.sma50x55BandUp
                                    : this.props.sma50x55BandDown,
                        },
                    },
                    {
                        tag: "Shapes",
                        key: "sma200x233Shading",
                        primitives: [
                            {
                                tag: "Polygon",
                                points: [
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma233),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma233),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma200),
                                    },
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma200),
                                    },
                                ],
                            },
                        ],
                        fillStyle: {
                            opacity: this.props.bandOpacity,
                            color:
                                newSma200 > newSma233
                                    ? this.props.sma200x233BandUp
                                    : this.props.sma200x233BandDown,
                        },
                    },
                    {
                        tag: "Shapes",
                        key: "sma365x377Shading",
                        primitives: [
                            {
                                tag: "Polygon",
                                points: [
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma377),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma377),
                                    },
                                    {
                                        x: du(d.index()),
                                        y: du(newSma365),
                                    },
                                    {
                                        x: du(d.index() - 1),
                                        y: du(oldSma365),
                                    },
                                ],
                            },
                        ],
                        fillStyle: {
                            opacity: this.props.bandOpacity,
                            color:
                                newSma365 > newSma377
                                    ? this.props.sma365x377BandUp
                                    : this.props.sma365x377BandDown,
                        },
                    },
                ],
            },
        };
    }
    filter(d, i) {
        return i > 0 && predef.filters.isNumber(d.sma20);
        return i > 0 && predef.filters.isNumber(d.sma50);
        return i > 0 && predef.filters.isNumber(d.sma200);
        return i > 0 && predef.filters.isNumber(d.sma365);
    }
}
/*  Export  Module  */
module.exports = {
    name: "buddhaSmaBands",
    description: "Buddha SMA Bands",
    calculator: buddhaSmaBands,
    tags: ["Buddha Indicators"],
    params: {
        period20: predef.paramSpecs.period(20),
        period21: predef.paramSpecs.period(21),
        sma20x21BandUp: predef.paramSpecs.color("#ff0000"),
        sma20x21BandDown: predef.paramSpecs.color("#800000"),

        period50: predef.paramSpecs.period(50),
        period55: predef.paramSpecs.period(55),
        sma50x55BandUp: predef.paramSpecs.color("#00FFFF"),
        sma50x55BandDown: predef.paramSpecs.color("#0000FF"),

        period200: predef.paramSpecs.period(200),
        period233: predef.paramSpecs.period(233),
        sma200x233BandUp: predef.paramSpecs.color("#FFFF00"),
        sma200x233BandDown: predef.paramSpecs.color("#FF7F00"),

        period365: predef.paramSpecs.period(365),
        period377: predef.paramSpecs.period(377),
        sma365x377BandUp: predef.paramSpecs.color("#FF00FF"),
        sma365x377BandDown: predef.paramSpecs.color("#800080"),

        bandOpacity: predef.paramSpecs.number(20, 10, 0),
    },
    plots: {
        sma20: { title: "20 SMA" },
        sma21: { title: "21 SMA" },
        sma50: { title: "50 SMA" },
        sma55: { title: "55 SMA" },
        sma200: { title: "200 SMA" },
        sma233: { title: "233 SMA" },
        sma365: { title: "365 SMA" },
        sma377: { title: "377 SMA" },
    },
    schemeStyles: {
        dark: {
            sma20: {
                color: "#FF0000",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma21: {
                color: "#800000",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma50: {
                color: "#00FFFF",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma55: {
                color: "#0000FF",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma200: {
                color: "#FFFF00",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma233: {
                color: "#FF7F00",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma365: {
                color: "#FF00FF",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
            sma377: {
                color: "#800080",
                lineStyle: 1,
                lineWidth: 1,
                opacity: 100,
            },
        },
    },
};
/*  END  CODE  */
